### PR TITLE
fix(ext/node): add ServerResponse#appendHeader

### PR DESCRIPTION
**Pull request by littledivy**
Fixes https://github.com/denoland/deno/issues/19993
Closes https://github.com/denoland/deno/issues/16679

### CODE REVIEW

**Review by satyarohith — APPROVED** [review]
LGTM

File: ext/node/polyfills/http.ts [hunk 1]
```diff
@@ -1419,6 +1419,20 @@ export class ServerResponse extends NodeWritable {
...
+      this.#hasNonStringHeaders = true;
+    }
+    if (this.#headers[name] === undefined) {
+      this.#headers[name] = value;
+    } else {
+      if (!Array.isArray(this.#headers[name])) {
+        this.#headers[name] = [this.#headers[name]];
+      }
+      this.#headers[name].push(value);
+    }
```
Comment (satyarohith) [inline ext/node/polyfills/http.ts:1433]:
```suggestion
    if (this.#headers[name] === undefined) {
      this.#headers[name] = value;
    } else {
      this.#headers[name] = [].concat(this.#headers[name], value);
    }
```

nit: trick if you to reduce if statements. `concat` returns a new array but in this case perf is not an issue considering header values are limited.

**Review by bartlomieju — COMMENTED** [review]
This method is already defined in `OutgoingMessage` class in `_http_outgoing.ts`. I don't think we should add another one, but instead we should rewrite `ServerResponse` to use ES5 classes and "mixin" both `OutgoingMessage` and `NodeWritable`

File: ext/node/polyfills/http.ts (hunk 1 above)
Comment (marvinhagemeister) [reply ext/node/polyfills/http.ts:1433]: If this API is used in a project using `.concat()` is an additional allocation per request though.

File: ext/node/polyfills/http.ts [hunk 2]
```diff
@@ -1419,6 +1419,20 @@ export class ServerResponse extends NodeWritable {
...
   }
+  appendHeader(name: string, value: string | string[]) {
+    if (Array.isArray(value)) {
+      this.#hasNonStringHeaders = true;
+    }
+    if (this.#headers[name] === undefined) {
+      this.#headers[name] = value;
+    } else {
+      if (!Array.isArray(this.#headers[name])) {
+        this.#headers[name] = [this.#headers[name]];
```
Comment (lucacasonato) [inline ext/node/polyfills/http.ts:1430]: You need to add a `this.#hasNonStringHeaders = true` here.

File: ext/node/polyfills/http.ts [hunk 3]
```diff
@@ -1419,6 +1419,20 @@ export class ServerResponse extends NodeWritable {
...
+      }
+      this.#headers[name].push(value);
```
Comment (lucacasonato) [inline ext/node/polyfills/http.ts:1432]: This creates an array like `["a", "b", ["c", "d"]]` (nested headers arrays). That doesn't actually work.

Please check what Node does here (whether it creates a new array that concats the two arrays, or whether it does this and then handles nested arrays in the serialization).

**Comment by lucacasonato** [comment]
@bartlomieju Their header handling is different unfortunately.

**Comment by bartlomieju** [comment]
@lucacasonato how so? If you look at `ServerResponse` class in Node.js: https://github.com/nodejs/node/blob/f2381bc7cb120e43bb97ffeccd19dfdc946c8b2c/lib/_http_server.js#L195-L230 you can see that it uses `OutgoingMessage` mixin to provide `appendHeader` method. Maybe our implementation got out of sync with what Node is doing?

**Comment by bartlomieju** [comment]
Discussed offline, it's fine to land as is, I'm gonna rewrite this module after https://github.com/denoland/deno/pull/24237 lands.

Please address Luca's comments before landing.

File: ext/node/polyfills/http.ts (hunk 3 above)
Comment (littledivy) [reply ext/node/polyfills/http.ts:1432]: This should work fine. I added a test. 

For  `appendHeader("foo", ["bar"])`,  `headers[name]` becomes `["bar"]` which is serialized properly when we set `#hasNonStringHeaders = true` (See L1422).

Nested headers array like `["foo", ["bar"]]` errors.

Comment (lucacasonato) [reply ext/node/polyfills/http.ts:1432]: I understand that toString will add spaces. However that doesn't work for things like Set-Cookie, where appending headers by joining them together with , is invalid.

We need to make sure this matches node behavior

Comment (littledivy) [reply ext/node/polyfills/http.ts]: Updated, similar to what Node is [doing](https://github.com/nodejs/node/blob/fbdfe9399cf6c660e67fd7d6ceabfb106e32d787/lib/_http_outgoing.js#L770-L782)

**Review by lucacasonato — APPROVED** [review]
LGTM!

**Pull request merged**